classes/local/store/object_file_system.php

```diff
@@ -737,7 +737,15 @@ protected function copy_from_external_to_local($contenthash) {
      */
     public function redirect_to_presigned_url($contenthash, $headers = array()) {
         try {
-            redirect($this->externalclient->generate_presigned_url($contenthash, $headers));
+            $signedurl = $this->externalclient->generate_presigned_url($contenthash, $headers);
+            // Set expires and cache-control values to match the presigned URL expiry, which may be
+            // different to values previously set.
+            header('Expires: '. gmdate('D, d M Y H:i:s', $signedurl->expiresat) .' GMT');
+            // Use 'private' to allow browser caching only, otherwise via a shared cache users might
+            // be able to redirect to content that was only supposed to be displayed to a different
+            // user.
+            header('Cache-Control: private, max-age=' . ($signedurl->expiresat - time()));
```

Comment:
This all looks great @sammarshallou 

There is still a part of me that wants to go the one extra step and do the full nuclear option where we completely unset all headers that have been set but not flushed by _pluginfile so that we only ever get the exact headers we do here. There are numerous ways of setting expires and it's really unclear to me how different browser might priorities conflicting headers.

ie loop over headers_list() and unset them. If headers_sent() already then we have to keep going but may log a debugging call

Reply:
OK! I have force-pushed a modified commit that uses header_remove to get rid of existing headers, and also adds a debugging message if headers were already sent. It took me 2 attempts because my first one used header_remove(null) but then I spotted the bit at the bottom of the manual page that says that only works in PHP 8, so I've changed it to loop around as suggested.

The only changes to previous commit are in that same section quoted above.